src/module_2/module_2.md

```diff
@@ -0,0 +1,3512 @@
+      <td>5.1</td>
+      <td>listerine</td>
+      <td>dental</td>
+      <td>[mouthwash]</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>2306493612164</td>
+      <td>1296e1e72f7f43ff28d7d285f880ad4d213fa8139233c7...</td>
+      <td>2020-06-21 15:09:26</td>
+      <td>2020-06-21</td>
+      <td>3</td>
+      <td>33667238658180</td>
+      <td>33667238658180</td>
+      <td>4.19</td>
+      <td>5.1</td>
+      <td>listerine</td>
+      <td>dental</td>
+      <td>[mouthwash]</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+print(f"Number of unique items in inventory: {df_inventory['variant_id'].nunique()}")
+print(f"Unique items in inventory: {len(df_inventory)}")
+print(f"Items sold merged with inventory: {len(orders_with_inventory)}")
+
+```
+
+    Number of unique items in inventory: 1733
+    Unique items in inventory: 1733
+    Items sold merged with inventory: 92361
+
+
+We see the value length of orders_with_inventory changes depending which type of merge we perform (left, inner, right, outer). This means that are product being sold that have no corresponde on inventory and products in inventory not being sold. Would be interesting to check those values and verify with Sales team what could be happening there:
+
+- Products in inventory not being sold might be a problem (OOS, wrong ids in one of the sides, horrible product that no one wants, etc)
+- Products being sold but not in inventory (this is more problematic because it is normal that old products are not on sale anymore, but everything being sold should be tracked and maintained in their internal DB, so will be good to check if we are assigning wrong ids when we sell them or maybe there is some issue in inventory ids)
+
+We could make some quick numbers on how many items are present whether in left or right hand
+
+
+```python
+left = df_orders_unrolled.merge(df_inventory, how='left',left_on='ordered_items', right_on='variant_id')
+inner = df_orders_unrolled.merge(df_inventory, how='inner',left_on='ordered_items', right_on='variant_id')
+right = df_orders_unrolled.merge(df_inventory, how='right',left_on='ordered_items', right_on='variant_id')
+outer = df_orders_unrolled.merge(df_inventory, how='outer',left_on='ordered_items', right_on='variant_id')
+
+print("All sold",len(left))
+print("Inner",len(inner))
+print("All inventory",len(right))
+print("All together",len(outer))
+
+
+```
+
+    All sold 107958
+    Inner 92361
+    All inventory 92617
+    All together 108214
+
+
+As a first step, what I'll do is perform an outer join as I did in the previous step, but enable indicator=True and filter for right_only. Those are the missing products in inventory which are being sold (we are assuming here that sales are contemporary, another reason of this mismatch might be old sales that we are reading in our df_orders, PIPOs, phase out products, etc). Anyway, we can also check that in the date of the order we're reading. First, we will do the mentioned one:
```

Comment:
I'm a little bit confused on why do you want to do this or what do you want to achieve. Maybe a cleaner way to count things in the left which are not in the right or viceversa is using sets:

```
items_in_orders = set(inspect_orders_with_inventory. ordered_items)
items_in_inventory = set(df_inventory. variant_id)
items_in_orders_only = len(items_in_orders - items_in_inventory)
```